List reconciliation

pkozlowski-opensource opened the pull request:
Draft: needs rebase and minor cleanups

AndrewKushnir commented on packages/core/src/render3/instructions/control_flow.ts:
```diff
@@ -150,6 +151,52 @@ export function ɵɵrepeaterCreate(
+    return this.trackByFn(index, lView[CONTEXT].$implicit);
+  }
+  override attach(index: number, lView: LView<RepeaterContext<unknown>>): void {
+    const dehydratedView =
+        findMatchingDehydratedView(this.lContainer, this.templateTNode.tView!.ssrId);
+    addLViewToLContainer(
+        this.lContainer, lView, index, shouldAddViewToDom(this.templateTNode, dehydratedView));
```
```suggestion
    const dehydratedView = lView[HYDRATION];
    addLViewToLContainer(
        this.lContainer, lView, index, shouldAddViewToDom(this.templateTNode, dehydratedView));
```

pkozlowski-opensource replied:
Thnx @AndrewKushnir - this was it!

alxhub commented on packages/core/src/render3/list_reconciliation.ts:
```diff
@@ -0,0 +1,295 @@
+ * an adapter interface that makes it possible to work with different internal data structures,
+ * regardless of the actual values of the incoming collection.
+ */
+export interface LiveCollection<T, V> {
```
Nit: you can use the abstract class as the interface - no need for a separate type.

alxhub commented on packages/core/src/render3/list_reconciliation.ts:
```diff
@@ -0,0 +1,295 @@
+ * The live collection reconciliation algorithm that perform various in-place operations, so it
+ * reflects the content of the new (incoming) collection.
+ *
+ * The reconciliation algorithm
```
Unfinished comment?

pkozlowski-opensource replied:
Yeh :-) Updated.

alxhub commented on packages/core/src/render3/list_reconciliation.ts:
```diff
@@ -0,0 +1,295 @@
+    let newEndIdx = newCollection.length - 1;
+
+    while (liveStartIdx <= liveEndIdx && liveStartIdx <= newEndIdx) {
+      // compare from the beginning
```
In general I think this type of algorithm is a really good case for descriptive comments - what state are we in? What just happened? What needs to happen next? etc.

alxhub commented on packages/core/test/render3/list_reconciliation_spec.ts:
```diff
@@ -0,0 +1,486 @@
+  }
+}
+
+class LoggingPrevCollection<T, V> extends LiveCollectionImpl<T, V> implements LiveCollection<T, V> {
```
```suggestion
class LoggingLiveCollection<T, V> extends LiveCollectionImpl<T, V> implements LiveCollection<T, V> {
```

pkozlowski-opensource replied:
+1, done!

alxhub commented on packages/core/src/render3/list_reconciliation.ts:
```diff
@@ -0,0 +1,295 @@
+        continue;
+      }
+
+      // compare from the end
```
I think I see a way of combining these algorithms to lower the code size here.

The fundamental observation is to note that lines 98-134 are the only place we need random (tail) access to the collection. Otherwise, we're only ever looking at the "next" item and could go through an iterator if needed.

I don't think we need to do this refactoring before landing this change, but we should talk about it.

pkozlowski-opensource replied:
Agreed. Will try different refactorings here but it would be great if we can do it post feature freeze. Let's chat / pair.

pkozlowski-opensource committed:
feat(core): add new list reconcilation algorithm

The new list reconcilation algorithm, an alternative to
the DefaultIterableListDiffer. It works by performing updates
in place instead of creating intermediate data describing changes
to apply. For lists expressed as an Array it performs additional
optimizations for the moves and swap scenarios.

The new list diffing approach is meant to be used in the new control
flow and should me much faster as compared to the ngFor with the
DefaultIterableListDiffer.

pkozlowski-opensource committed:
feat(core): the new list reconciliation algorithm for built-in for

This commit plugs the new list reconciliation into the new built-in repeater.

alxhub commented:
This PR was merged into the repository by commit 7d42dc3c023391e12ea607beb227fd4426e1694d.
